docs/advanced-config.md
> The official Splunk documentation for this page is [Configure the Java agent](https://help.splunk.com/en/splunk-observability-cloud/manage-data/available-data-sources/supported-integrations-in-splunk-observability-cloud/apm-instrumentation/instrument-a-java-application/configure-the-java-agent). For instructions on how to contribute to the docs, see [CONTRIBUTING.md](../CONTRIBUTING.md#documentation).

# Advanced configuration

The agent can be configured in the following ways:

* System property (example: `-Dotel.resource.attributes=service.name=my-java-app`)
* Environment variable (example: `export OTEL_RESOURCE_ATTRIBUTES=service.name=my-java-app`)

> System property values take priority over corresponding environment variables.

Below you will find all the configuration options supported by this distribution.

## Splunk configuration

| System property                         | Environment variable                    | Default value | Support      | Description                                                                                                                                                                                                                                                                                                                                                                                              |
|-----------------------------------------|-----------------------------------------|---------------|--------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `splunk.access.token`                   | `SPLUNK_ACCESS_TOKEN`                   | unset         | Stable       | (Optional) Auth token allowing exporters to communicate directly with the Splunk cloud, passed as `X-SF-TOKEN` header.                                                                                                                                                                                                                                                                                   |
| `splunk.realm`                          | `SPLUNK_REALM`                          | `none`        | Stable       | Specify your organization's realm name, such as `us0` or `us1. When the realm is set, telemetry data is sent directly to the ingest endpoint of Splunk Observability Cloud, bypassing the Splunk Distribution of the OpenTelemetry Collector. If no realm is specified, telemetry is routed to a Splunk OpenTelemetry Collector deployed locally. This configuration applies only to metrics and traces. |
| `splunk.metrics.force_full_commandline` | `SPLUNK_METRICS_FORCE_FULL_COMMANDLINE` | `false`       | Experimental | Adds the full command line as a resource attribute for all metrics. If false, commands longer than 255 characters are truncated.                                                                                                                                                                                                                                                                         |
| `splunk.trace-response-header.enabled`  | `SPLUNK_TRACE_RESPONSE_HEADER_ENABLED`  | `true`        | Stable       | Enables adding server trace information to HTTP response headers. See [this document](https://help.splunk.com/en/splunk-observability-cloud/manage-data/available-data-sources/supported-integrations-in-splunk-observability-cloud/apm-instrumentation/instrument-a-java-application/configure-the-java-agent#server-trace-information-0) for more information.                                         |

## Trace exporters

| System property                 | Environment variable            | Default value                    | Support | Description                                                                               |
|---------------------------------|---------------------------------|----------------------------------|---------|-------------------------------------------------------------------------------------------|
| `otel.exporter.otlp.endpoint`   | `OTEL_EXPORTER_OTLP_ENDPOINT`   | `http://localhost:4317`          | Stable  | The OTLP endpoint to connect to. Setting this will override the `splunk.realm` property.  |
| `otel.traces.exporter`          | `OTEL_TRACES_EXPORTER`          | `otlp`                           | Stable  | Select the traces exporter to use. We recommend using the default OTLP exporter (`otlp`)  |

The Splunk Distribution of OpenTelemetry Java uses the OTLP traces exporter as the default setting.

## Trace propagation configuration

| System property    | Environment variable | Default value          | Support | Description                                                                                                                                                                                     |
|--------------------|----------------------|------------------------|---------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `otel.propagators` | `OTEL_PROPAGATORS`   | `tracecontext,baggage` | Stable  | A comma-separated list of propagators that will be used. You can find the list of supported propagators [here](https://opentelemetry.io/docs/languages/java/configuration/#properties-general). |

If you wish to be compatible with older versions of the Splunk Distribution of OpenTelemetry Java (or the SignalFx
Tracing Java Agent) you can set the trace propagator to B3:

```bash
export OTEL_PROPAGATORS=b3multi
```

## Trace configuration

| System property                                                  | Environment variable                                             | Default value | Support | Description                                                                                                                                                                                                                                                                                                                                                                                                |
|------------------------------------------------------------------|------------------------------------------------------------------|---------------|---------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `otel.instrumentation.common.peer-service-mapping`               | `OTEL_INSTRUMENTATION_COMMON_PEER_SERVICE_MAPPING`               | unset         | Stable  | Used to add a `peer.service` attribute by specifying a comma separated list of mapping from hostnames or IP addresses. <details><summary>Example</summary>If set to `1.2.3.4=cats-service,dogs-service.serverlessapis.com=dogs-api`, requests to `1.2.3.4` will have a `peer.service` attribute of `cats-service` and requests to `dogs-service.serverlessapis.com` will have one of `dogs-api`.</details> |
| `otel.instrumentation.methods.include`                           | `OTEL_INSTRUMENTATION_METHODS_INCLUDE`                           | unset         | Stable  | Same as adding `@WithSpan` annotation functionality for the target method string. <details><summary>Format</summary>`my.package.MyClass1[method1,method2];my.package.MyClass2[method3]`</details>                                                                                                                                                                                                          |
| `otel.instrumentation.opentelemetry-annotations.exclude-methods` | `OTEL_INSTRUMENTATION_OPENTELEMETRY_ANNOTATIONS_EXCLUDE_METHODS` | unset         | Stable  | Suppress `@WithSpan` instrumentation for specific methods. <details><summary>Format</summary>`my.package.MyClass1[method1,method2];my.package.MyClass2[method3]`</details>                                                                                                                                                                                                                                 |
| `otel.resource.attributes`                                       | `OTEL_RESOURCE_ATTRIBUTES`                                       | unset         | Stable  | Comma-separated list of resource attributes added to every reported span. <details><summary>Example</summary>`key1=val1,key2=val2`</details>                                                                                                                                                                                                                                                               |
| `otel.service.name`                                              | `OTEL_SERVICE_NAME`                                              | unset         | Stable  | Sets the value of the `service.name` resource attribute. If `service.name` is also set in `otel.resource.attributes`, this setting takes precedence.                                                                                                                                                                                                                                                       |
| `otel.span.attribute.count.limit`                                | `OTEL_SPAN_ATTRIBUTE_COUNT_LIMIT`                                | unlimited     | Stable  | Maximum number of attributes per span.                                                                                                                                                                                                                                                                                                                                                                     |
| `otel.span.event.count.limit`                                    | `OTEL_SPAN_EVENT_COUNT_LIMIT`                                    | unlimited     | Stable  | Maximum number of events per span.                                                                                                                                                                                                                                                                                                                                                                         |
| `otel.span.link.count.limit`                                     | `OTEL_SPAN_LINK_COUNT_LIMIT`                                     | `1000`        | Stable  | Maximum number of links per span.                                                                                                                                                                                                                                                                                                                                                                          |

## Sampling configuration

| System property          | Environment variable     | Default value  | Support | Description |
| ------------------------ | ------------------------ | -------------- | ------- | ----------- |
| `otel.traces.sampler`    | `OTEL_TRACES_SAMPLER`     | `always_on`    | Stable  | The sampler to use for tracing.	|

Splunk Distribution of OpenTelemetry Java supports all standard samplers as provided by
[OpenTelemetry Java SDK Autoconfigure](https://opentelemetry.io/docs/languages/java/configuration/#properties-traces).
In addition, the distribution adds the following samplers:

### `internal_root_off`
Setting `otel.traces.sampler` to `internal_root_off` drops all traces with root spans where `spanKind` is `INTERNAL`, `CLIENT` or `PRODUCER`. This setting only keeps root spans where `spanKind` is `SERVER` and `CONSUMER`.

### `rules`
This sampler allows to ignore individual endpoints and drop all traces that originate from them.
It applies only to spans with `SERVER` kind.

For example, the following configuration results in all requests to `/healthcheck` to be excluded from monitoring:

```shell
export OTEL_TRACES_SAMPLER=rules
export OTEL_TRACES_SAMPLER_ARG=drop=/healthcheck;fallback=parentbased_always_on
```
All requests to downstream services that happen as a consequence of calling an excluded endpoint are also excluded.

The value of `OTEL_TRACES_SAMPLER_ARG` is interpreted as a semicolon-separated list of rules.
The following types of rules are supported:

- `drop=<value>`: The sampler drops a span if its `url.path` attribute has a substring equal to the provided value.
  You can provide as many `drop` rules as you want.
- `fallback=sampler`: Fallback sampler used if no `drop` rule matched a  given span.
  Supported fallback samplers are `always_on` and `parentbased_always_on`.
  Probability samplers such as `traceidratio` are not supported.

> If several `fallback` rules are provided, only the last one will be in effect.

If `OTEL_TRACES_SAMPLER_ARG` is not provided or has en empty value, no `drop` rules are configured and `parentbased_always_on` sampler is as default.

## Java agent configuration

| System property          | Environment variable     | Default value | Support | Description                                      |
|--------------------------|--------------------------|---------------|---------|--------------------------------------------------|
| `otel.javaagent.enabled` | `OTEL_JAVAAGENT_ENABLED` | `true`        | Stable  | Globally enables javaagent auto-instrumentation. |

## Profiler settings

| Setting                                   | Default                       | Description                                                                                                               |
|-------------------------------------------|-------------------------------|---------------------------------------------------------------------------------------------------------------------------|
| `splunk.profiler.enabled`                 | false                         | set to `true` to enable the profiler                                                                                      |
| `splunk.profiler.directory`               | system temp directory         | location of JFR files, defaults to `System.getProperty("java.io.tmpdir")`                                                 |
| `splunk.profiler.recording.duration`      | 20s                           | recording unit duration                                                                                                   |
| `splunk.profiler.keep-files`              | false                         | leave JFR files on disk if `true`                                                                                         |
| `splunk.profiler.logs-endpoint`           | http://localhost:4318/v1/logs | where to send OTLP logs, defaults to `otel.exporter.otlp.endpoint`                                                        |
| `splunk.profiler.call.stack.interval`     | 10000ms                       | how often to sample call stacks                                                                                           |
| `splunk.profiler.memory.enabled`          | false                         | set to `true` to enable all other memory profiling options unless explicitly disabled. Setting to `true` enables metrics. |
| `splunk.profiler.memory.event.rate`       | 150/s                         | allocation event rate                                                                                                     |
| `splunk.profiler.include.internal.stacks` | false                         | set to `true` to include stack traces of agent internal threads and stack traces with only JDK internal frames            |
| `splunk.profiler.tracing.stacks.only`     | false                         | set to `true` to include only stack traces that are linked to a span context                                              |
| `splunk.profiler.otlp.protocol`           | `http/protobuf`               | The transport protocol to use on profiling OTLP log requests. Options include `grpc` and `http/protobuf`.                 |

If the `splunk.profiler.enabled` option is not enabled, all profiling features are disabled. For
example, setting `splunk.profiler.memory.enabled` to `true` has no effect if
`splunk.profiler.enabled` is set to `false`. Similarly, there is no separate toggle for periodic
collection of call stacks (thread dumps), as this feature is enabled only when the profiler is
enabled.

Note: Setting `splunk.profiler.memory.enabled` to `true` automatically activates the splunk metrics.

## Other OpenTelemetry Java agent configuration

You can find all other Java agent configuration options
described [here](https://help.splunk.com/en/splunk-observability-cloud/manage-data/instrument-back-end-services/instrument-back-end-applications-to-send-spans-to-splunk-apm/instrument-a-java-application/configure-the-java-agent).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on
the [Splunk GDI specification](https://github.com/signalfx/gdi-specification/blob/v1.0.0/specification/repository.md),
and this repository adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased

## v2.26.1 - 2026-03-23

### General

- OpenTelemetry Instrumentation for Java has been updated to version 2.26.1.

### 🔒 Security fixes

- Fix unsafe deserialization in RMI instrumentation that could lead to remote code execution
  ([CVE-2026-33701](https://github.com/signalfx/splunk-otel-java/security/advisories/GHSA-h8w2-rv57-vc6f),
  [#16979](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/16979))

### ⚠️⚠️ Breaking changes ⚠️⚠️

- Context propagation format for RMI was changed [#16979](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/16979)

## v2.26.0 - 2026-03-18

### General

- OpenTelemetry Java SDK has been updated to version 1.60.1.
- OpenTelemetry Instrumentation for Java has been updated to version 2.26.0.

## v2.25.1 - 2026-02-26

### General

- Update CSA to version 26.2.1-1436.

## v2.25.0 - 2026-02-20

### General

- OpenTelemetry Java SDK has been updated to version 1.59.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.25.0.

## v2.24.0 - 2026-01-21

### General

- OpenTelemetry Java SDK has been updated to version 1.58.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.24.0.
- CSA version updated to 25.12.0-1409 [#2582](https://github.com/signalfx/splunk-otel-java/pull/2582).

## v2.23.0 - 2025-12-18

### General

- OpenTelemetry Java SDK has been updated to version 1.57.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.23.0.

## v2.22.0 - 2025-11-27

### General

- OpenTelemetry Java SDK has been updated to version 1.56.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.22.0.

## v2.21.1 - 2025-10-28

### General

- Update CSA version to 25.10.0-1399 

## v2.21.0 - 2025-10-22

### General

- OpenTelemetry Java SDK has been updated to version 1.55.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.21.0.

## v2.20.1 - 2025-09-22

### General

- OpenTelemetry Java SDK has been updated to version 1.54.1.
- OpenTelemetry Instrumentation for Java has been updated to version 2.20.1.

## v2.20.0 - 2025-09-17

### General

- OpenTelemetry Java SDK has been updated to version 1.54.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.20.0.

### 📈 Enhancements

- Add experimental context propagation for oracle db [#2434](https://github.com/signalfx/splunk-otel-java/pull/2434)

## v2.19.0 - 2025-08-21

### General

- OpenTelemetry Java SDK has been updated to version 1.53.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.19.0.

## v2.18.0 - 2025-07-23

Note: The 1.x version of `splunk-otel-java` has reached end of support.

### General

- OpenTelemetry Java SDK has been updated to version 1.52.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.18.1.

## v2.17.1 - 2025-07-11

### General

- OpenTelemetry Instrumentation for Java has been updated to version 2.17.1.

## v2.17.0 - 2025-06-28

### General

- OpenTelemetry Java SDK has been updated to version 1.51.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.17.0.

## v2.16.0 - 2025-05-21

### General

- OpenTelemetry Java SDK has been updated to version 1.50.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.16.0.

### ⚠️⚠️ Breaking changes ⚠️⚠️

- Remove operation name from graphql span name (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/13794)
- Remove deprecated property for disabling kafka metrics (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/13803)

## v2.15.0 - 2025-04-16

### General

- OpenTelemetry Java SDK has been updated to version 1.49.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.15.0.

## v2.14.0 - 2025-03-19

### General

- OpenTelemetry Java SDK has been updated to version 1.48.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.14.0.

## v2.13.1 - 2025-03-05

### General

- OpenTelemetry Instrumentation for Java has been updated to version 2.13.3.

## v2.13.0 - 2025-02-19

### General

- OpenTelemetry Java SDK has been updated to version 1.47.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.13.1.

## v2.12.0 - 2025-01-22

### General

- OpenTelemetry Java SDK has been updated to version 1.46.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.12.0.
- Rename `splunk.distro.version` to `telemetry.distro.version`.
- Declare compatibility with GDI spec 1.7.0 
  [#2148](https://github.com/signalfx/splunk-otel-java/issues/2148)

## v2.11.0 - 2025-01-09

### General

- OpenTelemetry Java SDK has been updated to version 1.45.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.11.0.

## v2.10.0 - 2024-11-22

### General

- OpenTelemetry Java SDK has been updated to version 1.44.1.
- OpenTelemetry Instrumentation for Java has been updated to version 2.10.0.

## v2.9.0 - 2024-10-23

### General

- OpenTelemetry Java SDK has been updated to version 1.43.0.
- OpenTelemetry Instrumentation for Java has been updated to version 2.9.0.

## v1.32.5 - 2024-09-20

### General

- Micrometer dependency has been updated to version 1.13.4.
- Protobuf dependency has been updated to version 3.25.5.

## v2.8.1 - 2024-09-20

### General

- Protobuf dependency has been updated to version 3.25.5.

## v2.8.0 - 2024-09-18

### General

- OpenTelemetry Java SDK has been updated to version 1.42.1.
- OpenTelemetry Instrumentation for Java has been updated to version 2.8.0.

## v1.32.4 - 2024-08-28

### General


profiler/src/main/java/com/splunk/opentelemetry/profiler/Configuration.java
/*
 * Copyright Splunk Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.splunk.opentelemetry.profiler;

import static com.splunk.opentelemetry.SplunkConfiguration.PROFILER_ENABLED_PROPERTY;
import static java.util.logging.Level.WARNING;

import com.splunk.opentelemetry.SplunkConfiguration;
import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
import java.time.Duration;
import java.util.logging.Logger;
import javax.annotation.Nullable;

public class Configuration {
  private static final Logger logger = Logger.getLogger(Configuration.class.getName());

  static final String CONFIG_KEY_OTLP_PROTOCOL = "otel.exporter.otlp.protocol";
  static final String CONFIG_KEY_OTEL_OTLP_URL = "otel.exporter.otlp.endpoint";

  /* Keys visible for testing */
  static final String CONFIG_KEY_PROFILER_DIRECTORY = "splunk.profiler.directory";
  static final String CONFIG_KEY_RECORDING_DURATION = "splunk.profiler.recording.duration";
  static final String CONFIG_KEY_KEEP_FILES = "splunk.profiler.keep-files";
  static final String CONFIG_KEY_INGEST_URL = "splunk.profiler.logs-endpoint";
  static final String CONFIG_KEY_PROFILER_OTLP_PROTOCOL = "splunk.profiler.otlp.protocol";
  static final String CONFIG_KEY_MEMORY_ENABLED = "splunk.profiler.memory.enabled";
  static final String CONFIG_KEY_MEMORY_EVENT_RATE_LIMIT_ENABLED =
      "splunk.profiler.memory.event.rate-limit.enabled";
  static final String CONFIG_KEY_MEMORY_EVENT_RATE = "splunk.profiler.memory.event.rate";
  static final String CONFIG_KEY_MEMORY_NATIVE_SAMPLING = "splunk.profiler.memory.native.sampling";
  static final String CONFIG_KEY_CALL_STACK_INTERVAL = "splunk.profiler.call.stack.interval";
  static final String CONFIG_KEY_INCLUDE_AGENT_INTERNALS =
      "splunk.profiler.include.agent.internals";
  // Include stacks where every frame starts with jvm/sun/jdk
  static final String CONFIG_KEY_INCLUDE_JVM_INTERNALS = "splunk.profiler.include.jvm.internals";
  static final String CONFIG_KEY_INCLUDE_INTERNAL_STACKS =
      "splunk.profiler.include.internal.stacks";
  static final String CONFIG_KEY_TRACING_STACKS_ONLY = "splunk.profiler.tracing.stacks.only";
  static final String CONFIG_KEY_STACK_DEPTH = "splunk.profiler.max.stack.depth";

  private static final boolean HAS_OBJECT_ALLOCATION_SAMPLE_EVENT = getJavaVersion() >= 16;

  private static final String DEFAULT_PROFILER_DIRECTORY = System.getProperty("java.io.tmpdir");
  private static final Duration DEFAULT_RECORDING_DURATION = Duration.ofSeconds(20);
  private static final Duration DEFAULT_CALL_STACK_INTERVAL = Duration.ofSeconds(10);

  public static void log(ConfigProperties config) {
    logger.info("-----------------------");
    logger.info("Profiler configuration:");
    log(PROFILER_ENABLED_PROPERTY, SplunkConfiguration.isProfilerEnabled(config));
    log(CONFIG_KEY_PROFILER_DIRECTORY, getProfilerDirectory(config));
    log(CONFIG_KEY_RECORDING_DURATION, getRecordingDuration(config).toMillis() + "ms");
    log(CONFIG_KEY_KEEP_FILES, getKeepFiles(config));
    log(CONFIG_KEY_PROFILER_OTLP_PROTOCOL, getOtlpProtocol(config));
    log(CONFIG_KEY_INGEST_URL, getIngestUrl(config));
    log(CONFIG_KEY_OTEL_OTLP_URL, config.getString(CONFIG_KEY_OTEL_OTLP_URL));
    log(CONFIG_KEY_MEMORY_ENABLED, getMemoryEnabled(config));
    if (getMemoryEventRateLimitEnabled(config)) {
      log(CONFIG_KEY_MEMORY_EVENT_RATE, getMemoryEventRate(config));
    }
    log(CONFIG_KEY_MEMORY_NATIVE_SAMPLING, getUseAllocationSampleEvent(config));
    log(CONFIG_KEY_CALL_STACK_INTERVAL, getCallStackInterval(config).toMillis() + "ms");
    log(CONFIG_KEY_INCLUDE_AGENT_INTERNALS, getIncludeAgentInternalStacks(config));
    log(CONFIG_KEY_INCLUDE_JVM_INTERNALS, getIncludeJvmInternalStacks(config));
    log(CONFIG_KEY_TRACING_STACKS_ONLY, getTracingStacksOnly(config));
    log(CONFIG_KEY_STACK_DEPTH, getStackDepth(config));
    logger.info("-----------------------");
  }

  private static void log(String key, @Nullable Object value) {
    logger.info(String.format("%39s : %s", key, value));
  }

  public static String getIngestUrl(ConfigProperties config) {
    String ingestUrl = config.getString(CONFIG_KEY_INGEST_URL);

    if (ingestUrl == null) {
      String defaultIngestUrl = getDefaultLogsEndpoint(config);
      ingestUrl = config.getString(CONFIG_KEY_OTEL_OTLP_URL, defaultIngestUrl);

      if (ingestUrl.startsWith("https://ingest.") && ingestUrl.endsWith(".signalfx.com")) {
        logger.log(
            WARNING,
            "Profiling data can not be sent to {0}, using {1} instead. "
                + "You can override it by setting "
                + CONFIG_KEY_INGEST_URL,
            new Object[] {ingestUrl, defaultIngestUrl});
        return defaultIngestUrl;
      }

      if ("http/protobuf".equals(getOtlpProtocol(config))) {
        ingestUrl = maybeAppendHttpPath(ingestUrl);
      }
    }

    return ingestUrl;
  }

  private static String maybeAppendHttpPath(String ingestUrl) {
    if (!ingestUrl.endsWith("v1/logs")) {
      if (!ingestUrl.endsWith("/")) {
        ingestUrl += "/";
      }
      ingestUrl += "v1/logs";
    }

    return ingestUrl;
  }

  private static String getDefaultLogsEndpoint(ConfigProperties config) {
    return "http/protobuf".equals(getOtlpProtocol(config))
        ? "http://localhost:4318/v1/logs"
        : "http://localhost:4317";
  }

  public static String getOtlpProtocol(ConfigProperties config) {
    return config.getString(
        CONFIG_KEY_PROFILER_OTLP_PROTOCOL,
        config.getString(CONFIG_KEY_OTLP_PROTOCOL, "http/protobuf"));
  }

  public static boolean getMemoryEnabled(ConfigProperties config) {
    return config.getBoolean(CONFIG_KEY_MEMORY_ENABLED, false);
  }

  public static boolean getMemoryEventRateLimitEnabled(ConfigProperties config) {
    return config.getBoolean(CONFIG_KEY_MEMORY_EVENT_RATE_LIMIT_ENABLED, true);
  }

  public static String getMemoryEventRate(ConfigProperties config) {
    return config.getString(CONFIG_KEY_MEMORY_EVENT_RATE, "150/s");
  }

  public static boolean getUseAllocationSampleEvent(ConfigProperties config) {
    // Using jdk16+ ObjectAllocationSample event is disabled by default
    return HAS_OBJECT_ALLOCATION_SAMPLE_EVENT
        && config.getBoolean(CONFIG_KEY_MEMORY_NATIVE_SAMPLING, false);
  }

  public static Duration getCallStackInterval(ConfigProperties config) {
    return config.getDuration(CONFIG_KEY_CALL_STACK_INTERVAL, DEFAULT_CALL_STACK_INTERVAL);
  }

  public static boolean getIncludeAgentInternalStacks(ConfigProperties config) {
    boolean includeInternals = config.getBoolean(CONFIG_KEY_INCLUDE_INTERNAL_STACKS, false);
    return config.getBoolean(CONFIG_KEY_INCLUDE_AGENT_INTERNALS, includeInternals);
  }

  public static boolean getIncludeJvmInternalStacks(ConfigProperties config) {
    boolean includeInternals = config.getBoolean(CONFIG_KEY_INCLUDE_INTERNAL_STACKS, false);
    return config.getBoolean(CONFIG_KEY_INCLUDE_JVM_INTERNALS, includeInternals);
  }

  public static boolean getTracingStacksOnly(ConfigProperties config) {
    return config.getBoolean(CONFIG_KEY_TRACING_STACKS_ONLY, false);
  }

  public static int getStackDepth(ConfigProperties config) {
    return config.getInt(CONFIG_KEY_STACK_DEPTH, 1024);
  }

  public static boolean getKeepFiles(ConfigProperties config) {
    return config.getBoolean(CONFIG_KEY_KEEP_FILES, false);
  }

  public static String getProfilerDirectory(ConfigProperties config) {
    return config.getString(CONFIG_KEY_PROFILER_DIRECTORY, DEFAULT_PROFILER_DIRECTORY);
  }

  public static Duration getRecordingDuration(ConfigProperties config) {
    return config.getDuration(CONFIG_KEY_RECORDING_DURATION, DEFAULT_RECORDING_DURATION);
  }

  private static int getJavaVersion() {
    String javaSpecVersion = System.getProperty("java.specification.version");
    if ("1.8".equals(javaSpecVersion)) {
      return 8;
    }
    return Integer.parseInt(javaSpecVersion);
  }
}
